File: pe-lwjgl-binding/src/main/java/org/saar/lwjgl/opengl/objects/ClipPlane.java
package org.saar.lwjgl.opengl.objects;

import org.joml.Vector4f;
import org.joml.Vector4fc;
import org.saar.maths.utils.Vector4;

public class ClipPlane {

    public static final ClipPlane NONE = ClipPlane.of(0, 0, 0, 0);

    private final Vector4fc value;

    private ClipPlane(Vector4f value) {
        this.value = value;
    }

    public static ClipPlane of(Vector4fc value) {
        return new ClipPlane(Vector4.of(value).normalize());
    }

    public static ClipPlane of(float a, float b, float c, float d) {
        return new ClipPlane(Vector4.of(a, b, c, d).normalize());
    }

    public static ClipPlane ofAbove(float height) {
        return ClipPlane.of(0, +1, 0, -height);
    }

    public static ClipPlane ofBelow(float height) {
        return ClipPlane.of(0, -1, 0, +height);
    }

    public Vector4fc getValue() {
        return this.value;
    }
}

File: pe-lwjgl-binding/src/main/java/org/saar/lwjgl/opengl/clipplane/ClipPlane.kt
package org.saar.lwjgl.opengl.clipplane

import org.joml.Vector4fc
import org.saar.maths.utils.Vector4

class ClipPlane private constructor(val value: Vector4fc) {

    companion object {
        @JvmStatic
        val NONE = of(0f, 0f, 0f, 0f)

        @JvmStatic
        fun of(value: Vector4fc) = ClipPlane(value.normalize(Vector4.create()))

        @JvmStatic
        fun of(a: Float, b: Float, c: Float, d: Float) = ClipPlane(Vector4.of(a, b, c, d).normalize())

        @JvmStatic
        fun ofAbove(height: Float) = of(0f, +1f, 0f, -height)

        @JvmStatic
        fun ofBelow(height: Float) = of(0f, -1f, 0f, +height)
    }


    override fun equals(other: Any?) = other is ClipPlane && other.value == this.value

    override fun hashCode() = this.value.hashCode()

    override fun toString() = "ClipPlane[${this.value.toString()}]"
}